create initial outline of it maintainance policy (#126)
carpentries/trainers: policy/it_curriculum_maintainance.md

# Instructor Training Curriculum Maintainance

This document describes how the Carpentries Community maintains the instructor training curriculum.  
Because this curriculum is taught by a select group of individuals and serves as an entrypoint for 
instructors who teach other curricula for The Carpentries, maintainance of the IT curriculum involves
a greater number of stakeholders.  

The IT curriculum does not have a standard Curriculum Advisory Committee, there is the whole trainer 
community and their elected leadership panel. 

> 

Reviewer: I'm always +1 for more explicit definitions.

This comment hints that IT curriculum re-writes are common. Is that the actual plan? or are the more recent IT changes the biggest re-write planning for now?
If there aren't plans for another large re-write of the materials, I think the issues/prs/edits are on par with any of the other core lessons.

> 

Author: IT does have a "major revision" goal of approximately annually, some are just clearing complex issues, but this curriculum does change more than other core lessons as far as I know. This lesson aims to keep up-to-date-ish with educational psycholgy research that's relevant, where other lessons stay up to date with libraries/packages/tools.  Research, conceptually, moves faster than tools, I think? @karenword and @ErinBecker may be able to provide more insight on this matter.

> 

Author: definitions added

> 

Author: ```suggestion
## Role of Maintainers 

- certify as maintainers for the carpentries 
- (regular maintainer duties, via pointer?) 
- refer issues for trainer discussion
- provide consistent peer review during major updates 
```

## Maintainer Selection

In addition to the Core Team IT Team, there should be at least X trainers as maintainers. 

> 

Reviewer: this is something I grapple with while planning things out for the larger maintainer community.
2 seems a bit low, but the other lessons do not have the Core Team IT Team on maintenance duties as well.
So this seems okay.

Since the IT sessions are planned in advance, we can be a bit more explicit about clearing all the issues before the next round of training (or at least have a plan for them). That way people have more concrete deadlines and can work around that. (maybe that's something I should think about for the broader maintainer community as well.... )

> 

Author: I think we don't need to be so specific about the technical things because we *can* ask infrastructure if needed? is that true? @fmichonneau or should we be explicit here that one person needs to be able to do that stuff on this repo. 

that issue curation role of submitting issues is interesting and likely super helpful, but currently not what maintainers are doing... @karenword @ jt14den thoughts on how to handle this role practically?

> 

Reviewer 2: I really prefer to have issues submitted by people who originate ideas and often ask people to do this when things come up in meetings. However, a) this doesn't always happen and b) there are sometimes broader patterns that aren't just one person's idea. I'm realizing that there are things on my priority list that don't have issues at the moment, so I have to admit there would be a value in having someone with the designated responsibility to be sure that problems turn into issues.

> 

Author: I created an issue to preserve part of this discussion so that we can get this policy closed. Let's keep the minimum at 3 for now and see how it goes.

> One member of the leadership panel should participate as a maintainer and, with the Core Team
> faciliate updates between the two groups. 

Author: ```suggestion
One member of the leadership panel should participate as a maintainer and, with the Core Team 
Instructor Training Team facilitate information sharing between the maintainers and leadership panel. 
```

Reviewer: does this mean we are only looking for 1 maintainer from the community? (assuming x = 2 maintainers)

Author: at a minimu, yes, but can go above if needed

> 

Author: there have always been more maintainers beyond the Core team. I'll add goals above.

## Changes requiring input outside of maintainers

> what gets referred to leadership,
> what role the whole trainer community has in maintenance

- Minor changes can be merged by any maintainer. 
- Changes to policy must be referred to the trainer leadership panel
- The Core team curriculum team supports the list of starting points
- Maintainers may request using a trainer meeting to discuss proposals to the currciulum that 
  represent a significant change. 
- organizational and structural changes follow the major update process below

> 

Reviewer 3: Let's borrow from semantic versioning here (https://semver.org/). 

* Patches are typo fixes and in situ rearticulations of paragraphs that just don't make sense. There's no change in lesson or in-episode order and there is no expectation that notice needs to be given to trainers preparing to run a workshop. (I.e. if they print out and highlight a lesson and a patch is applied, there will be no unpleasant surprises.) No restriction on patch application, either in time or by person. No existing issue is needed.
* Minor updates are PRs from issues. They are changes which should be locked out (x weeks) before a training is happening -- or trainers should be expected to fork instructor training to avoid minor updates. This may make narrow windows where minor updates can be applied, but that's totally OK. So long as each minor update has an issue associated with it, people can follow the conversation and intention and commits as part of their instruction prep. 
* Major updates are the yearly maintenance that needs to be in its own branch and has all of the normal consultations. I'm... pretty against how we did it this year as the cognitive load of a different org or different major repo or... whatever... just bounced off my brain. We'll probably want to get @fmichonneau to weigh in on something like https://tigeroakes.com/posts/til-github-pages-two-branches-github-actions/ so we have a "training/proposed-changes/" site but... that's outside this specific discussion. Just dropping it here so we don't forget. 

As an aside, we probably want to turn on bors https://bors.tech/ for this repo to better handle text-merging in PRs if we go down this route. I can attest to the utility of it from my own projects.

> 

Author: we probably can't lock it before every training, because I think they're mostish weeks? @maneesha @karenword

> 

Author: I've incorporated most of these definitions above, but welceom feedback

> 

Reviewer 3: Ah, I suggested forking by instructors in comment above. SArah, if you think that's reasonable, merge that convo in here? If we say "The moment your site goes up, fork the repo" and have that be part of prep, the constraints on updates are nicely relaxed and provides for better scale, at the cost of higher cognitive load during prep.

> 

Reviewer 3: Regarding versions, and thinking back to this convo, maybe move policy discussion out of versions, or make it a specific kind of major udpate? Its ordering suggested that it was a version update and broke me from thinking of this as a semver based system

> 

Reviewer 2: Yeah, locking out changes prior to instruction is probably not a sustainable move as we scale. However, we could update our one-month email to Trainers, for example, to advise them to fork the repo if they are concerned about interference from minor updates. Or advise them to check out active PRs... we could have a wait time before PRs are merged, so they would know what was coming down the pipeline with no surprises?

> 

Reviewer 3: Absolutely recommend that they fork the repo by *default*. We can work up some instructions for merging from main into their fork, but if you folks know that you're starting from a place of not mucking people up -- that will make life so much easier. 

Basically: "You're going to need to prep something, and there will always be some change somewhere. Fork now, and teach from that -- it's good enough. If you feel bored on the plane, look at the last month of commits for the sections you're going to teach."

> 

Author: github has main upstream fetch really easy recently, so that's easy and I put in a [PR on the forking](https://github.com/carpentries/instructor-training/pull/1365) to the guide (maybe it goes in handbook? idk)

# Instructor Training Curriculum Maintainance

This document describes how the Carpentries Community maintains the instructor training curriculum.  
Because this curriculum is taught by a select group of individuals and serves as an entrypoint for 
instructors who teach other curricula for The Carpentries, maintainance of the IT curriculum involves
a greater number of stakeholders.  

The IT curriculum does not have a standard Curriculum Advisory Committee, there is the whole trainer 
community and their elected leadership panel. 

## Role of Maintainers 

- certify as maintainers for the carpentries 

> - (regular maintainer duties, via pointer?) 

Author: that page is super helpful! 
can you also link the current guildelines (if they hav ea permanent home)?

**Major updates** are updates that make multiple conceptual changes or change the key points of episodes. This includes PRs that resolve multiple interdependent issues; represent significant changes in terminology or framing; change which topics are covered or substantially reorder material. 

## Role of Maintainers 

- perform routine maintenance duties as [outlined by the Maintainer community](https://carpentries.github.io/maintainer-onboarding/05-communicate-advisors/index.html)
- refer issues to Trainer community for discussion
- provide peer review during major updates (see below)
- recruit and delegate to individuals or focus groups to collaborate on specific topics or tasks of interest
- complete certification as a Carpentries Maintainer as possible and practical

PR Handling:
- Patches may be merged by any maintainer without discussion. 
- Minor changes can be merged by any maintainer who did not contribute to the PR or after discussion among maintainers.

> - Changes to policy must be approved by the trainer leadership panel and linked to a proposal in that group. 

Reviewer 3: policy and/or major updates? Language needs to be consistent. is it a change,update, or revision?

Reviewer 3: Argh, ok, skipping around. This section needs to be reorganised. Put all three (Four?) levels of acceptance below with the major protocol. I didn't realise that we're intentionally skipping major.

- The Core Team curriculum team supports the list of starting points for teaching demonstrations. 

> - Maintainers may request using a trainer meeting to discuss proposals to the currciulum that 

Reviewer 3: This should be its own heading under line 49, and should be a general: "Any instructor trainer can ask for a pause on an issue if they think it deserves discussion at the next meeting. After the next meeting, the pause is immediately dropped unless there are strong reasons to the contrary.

Reviewer 2: I like the idea of having a section in special maintenance procedures about using Trainer meetings. This is something I'd like to do more often, not only for discussion but also for active brainstorming & team work on focused topics. Guidelines and structure could help.

## Maintainer Selection

This group should consist of at least 3 people:
- 1 designated member of The Carpentries Core Team
- 1 member of Trainers Leadership
- 1 additional Trainer

At least one member of this group should be certified as a Carpentries Maintainer, trainers
may be appointed prior to certification, but will take steps to complete that certification
upon appointment.  The Trainers leadership shall approve appointments to the maintainer 
committee each year. Additional maintainers may be added at any time by recommendation of
the existing maintainers with approval of the leadership panel. 

## Special Maintance Procedures/Sections

### Major updates and releases

To initiate a major update, the maintainers will submit a proposal for approval to the leadership panel. 
  - technical workflow (after consultation with the Core Team Infratrastructure team)
  - scope
  - timeline for notifiying the trainer community at large
  - how the process addressess challenges faced in previous major updates

The process must include:
  - time for pilot teaching by trainers that are not maintainers, members of the leadership panel, or core team

>   - at least two weeks of asynchronous 

Reviewer 4: Missing something at the end - asynchronous review? feedback?

Author: yes, changed now

>   - at least two weeks of asynchronous 

Author: ```suggestion
  - at least two weeks of asynchronous feedback with notifications for all trainers
```

Before a major update is merged, the maintainers submit a proposal to the leadership panel that describes:
  - a summary of changes
  - the completed opportunities for community feedback
  - opportunities to improve the major update process for next time.  

>   

Reviewer 3: And discussion of forking and how to deal with TTTs scheduled for the update period. I still remember the frustration that caused me to make the etherpad exporter.

PErhaps we want to codify the reverse, suggesting that all instructors preparing to teach fork and teach from their fork?

Author: This is a good point. In what's written here I'm additionally trying to balance leaving flexibility and not codifying technologically dependent things as much as possible so that we don't for example, *have* to update this if we find better workflows, but that we do have to require communication to support trainers.

### Starting Points for Teaching Demonstrations

This page is maintained by the Curriculum team. 
Approved starting points shall be at the start of an episode and avoid dependencies. 

> Recommended changes to this page will be referred to the curriuculum team for approval. 

Reviewer 5: ```suggestion
Recommended changes to this page will be referred to the curriculum team for approval. 
```

This group shall meet a minimum of 4 times per year to approve new starting points. 

> Proposals for new starting points in issues will be tagged by the maintainers for consideration by this group. 

Reviewer 5: ```suggestion
Proposals for new starting points in issues will be tagged by the maintainers for consideration by this group. 

When a new official curriculum is introduced, the Curriculum Team will prepare a list of starting points to be added.
In this circumstance, a member of the Curriculum Team will open a pull request to add the new starting points and tag the Instructor Training Maintenance Team for discussion and approval.
```

## Maintainer Selection

This group should consist of at least 3 people:
- 1 designated member of The Carpentries Core Team
- 1 member of Trainers Leadership
- 1 additional Trainer

At least one member of this group should be certified as a Carpentries Maintainer, trainers
may be appointed prior to certification, but will take steps to complete that certification
upon appointment.  The Trainers leadership shall approve appointments to the maintainer 
committee each year. Additional maintainers may be added at any time by recommendation of
the existing maintainers with approval of the leadership panel. 

## Special Maintance Procedures/Sections

### Major updates and releases

To initiate a major update, the maintainers will submit a proposal for approval to the leadership panel. 
  - technical workflow (after consultation with the Core Team Infratrastructure team)
  - scope
  - timeline for notifiying the trainer community at large
  - how the process addressess challenges faced in previous major updates

The process must include:
  - time for pilot teaching by trainers that are not maintainers, members of the leadership panel, or core team

>   - at least two weeks of asynchronous feedback with notifications for all trainers

Reviewer 2: In this last update we had pilot -> feedback -> changes in response to feedback -> final commenting period with survey (in which everyone was pretty tired of giving feedback). I'm not sure how much of this should be required, just pointing out that pilot and feedback aren't necessarily the end of the story.

Author: I think I like leaving the 2 as the minimum because a change that was big enough but less intense than this last one might not have as many changes after feedback and we may want option.  This requirement is that there will be a process outlined for each one so it can be more.

Before a major update is merged, the maintainers submit a proposal to the leadership panel that describes:
  - a summary of changes
  - the completed opportunities for community feedback
  - opportunities to improve the major update process for next time.  
  
### Starting Points for Teaching Demonstrations

This page is maintained by the Curriculum team. 
Approved starting points shall be at the start of an episode and avoid dependencies. 
Recommended changes to this page will be referred to the curriuculum team for approval. 

> This group shall meet a minimum of 4 times per year to approve new starting points. 

Reviewer 2: Who is "This group" here -- the curriculum team?

Reviewer 5: Not entirely sure - I think the group being referred to is the Trainer Leadership, and the wording should be changed to "meet a minimum of four timers per year to recommend new starting points." But TBH I think the most likely route for new starting points is directly from the Curriculum Team, because it is fair to expect us to be more aware of new and changing curriculum than other members of the Trainer community.

Author: I think I fixed this, please re-read?

Reviewer 5: Looks good to me 👍

- Organizational and structural changes follow the major update process below. 

## Maintainer Selection

This group should consist of at least 3 people:
- 1 designated member of The Carpentries Core Team
- 1 member of Trainers Leadership
- 1 additional Trainer

At least one member of this group should be certified as a Carpentries Maintainer, trainers
may be appointed prior to certification, but will take steps to complete that certification
upon appointment.  The Trainers leadership shall approve appointments to the maintainer 
committee each year. Additional maintainers may be added at any time by recommendation of
the existing maintainers with approval of the leadership panel. 

## Special Maintance Procedures/Sections

### Major updates and releases

To initiate a major update, the maintainers will submit a proposal for approval to the leadership panel. 
  - technical workflow (after consultation with the Core Team Infratrastructure team)
  - scope
  - timeline for notifiying the trainer community at large
  - how the process addressess challenges faced in previous major updates

> The process must include:

Author: ```suggestion
The process must include, but is not limited to:
```

**Major updates** are updates that make multiple conceptual changes or change the key points of episodes. This includes PRs that resolve multiple interdependent issues; represent significant changes in terminology or framing; change which topics are covered or substantially reorder material. 

## Role of Maintainers 

- perform routine maintenance duties as [outlined by the Maintainer community](https://carpentries.github.io/maintainer-onboarding/05-communicate-advisors/index.html)
- refer issues to Trainer community for discussion
- provide peer review during major updates (see below)
- recruit and delegate to individuals or focus groups to collaborate on specific topics or tasks of interest
- complete certification as a Carpentries Maintainer as possible and practical

PR Handling:
- Patches may be merged by any maintainer without discussion. 
- Minor changes can be merged by any maintainer who did not contribute to the PR, after approval by another trainer, or after discussion among maintainers.

> - PRs for major major changes or that impact polciy are described in the special sections below. 

Reviewer 4: ```suggestion
- PRs for major changes or that impact policy are described in the special sections below. 
```

## Maintainer Selection

This group should consist of at least 3 people:
- 1 designated member of The Carpentries Core Team
- 1 member of Trainers Leadership
- 1 additional Trainer

At least one member of this group should be certified as a Carpentries Maintainer, trainers

> may be appointed prior to certification, but will take steps to complete that certification

Reviewer 4: ```suggestion
may be appointed prior to certification, but must take steps to complete that certification
```

> upon appointment.  The Trainers leadership shall approve appointments to the maintainer 

Reviewer 4: ```suggestion
upon appointment.  The Trainers Leadership shall approve appointments to the maintainer 
```

committee each year. Additional maintainers may be added at any time by recommendation of

> the existing maintainers with approval of the leadership panel. 

Reviewer 4: ```suggestion
the existing maintainers with approval of the Leadership Panel. 
```

## Special Maintance Procedures/Sections

### Major Updates and Releases

> To initiate a major update, the maintainers will submit a proposal for approval to the leadership panel. 

Reviewer 4: ```suggestion
To initiate a major update, the maintainers will submit a proposal for approval to the Leadership Panel. 
```

>   - technical workflow (after consultation with the Core Team Infratrastructure team)

Reviewer 4: ```suggestion
  - technical workflow (after consultation with the Core Team Infrastructure team)
```

  - scope
  - timeline for notifiying the trainer community at large
  - how the process addressess challenges faced in previous major updates

The process must include, but is not limited to:

>   - time for pilot teaching by trainers that are not maintainers, members of the leadership panel, or core team

Reviewer 4: ```suggestion
  - time for pilot teaching by trainers that are not maintainers, members of the leadership panel, or Core Team
```

  - at least two weeks of asynchronous feedback with notifications for all trainers

> Before a major update is merged, the maintainers submit a proposal to the leadership panel that describes:

Reviewer 4: ```suggestion
Before a major update is merged, the maintainers submit a proposal to the Leadership Panel that describes:
```

  - a summary of changes
  - the completed opportunities for community feedback
  - opportunities to improve the major update process for next time.  

> After a Major Update, a release will be issued in collaboration iwth the Core Team Infrastructure Team. 

Reviewer 4: ```suggestion
After a Major Update, a release will be issued in collaboration with the Core Team Infrastructure Team. 
```

Instructor training related procedure and policy will go through the Trainer Leadership Panel. 
The PR to the curriculum repository should be linked to a proposal by that group linking the panel
approval to these changes. 

The "Suggested Lessons for Teaching Demonstrations" page will be maintained by the Curriculum team. 
Approved starting points shall be at the start of an episode and avoid dependencies. 
Recommended changes to this page will be referred to the curriculum team for approval. 
Proposals for new starting points in issues will be tagged by the maintainers for consideration by this group. 

When a new official curriculum is introduced, the Curriculum Team will prepare a list of starting points to be added.
In this circumstance, a member of the Curriculum Team will open a pull request to add the new starting points and tag the Instructor Training Maintenance Team for review and approval.

### Using Trainer Meetings 

Any instructor trainer may ask for a pause on a Pull Request if they think it deserves discussion at the next meeting. After the next meeting, the pause is immediately dropped unless there are strong reasons to the contrary.

> ### Maintainer Notes
> 
> how/where additional information is documented (ie addresses IT/1173)

Author: ```suggestion
```
I can't think of anything better than that issue for now, so I think we should just leave the issue and approve this.

Author: Once this is committed, then this is ready for approval and merging.
